modules/server/src/main/java/org/jboss/wsf/stack/cxf/security/authentication/SubjectCreator.java

```diff
@@ -97,12 +103,13 @@ public Subject createSubject(SecurityDomainContext ctx, String name, String pass
          }
          if (isDigest && created != null && nonce != null) { // username token profile is using digest
             // verify client's digest
-            ClearPassword clearPassword = identity.getCredential(PasswordCredential.class).getPassword(ClearPassword.class);
-            // only realms supporting getCredential with clear password can be used with Username Token profile
-            if (clearPassword == null) {
+            TwoWayPassword recoveredTwoWayPassword = identity.getCredential(PasswordCredential.class).getPassword(TwoWayPassword.class);
```

Comment:
Can this `identity.getCredential(PasswordCredential.class)` be another password type like `DigestPassword` instead of `TwoWayPassword` ?  
If it isn't a TwoWayPassword type, this will return a null object and following `recoveredTwoWayPassword.getAlgorithm()` throws a NPE.  Will this happen ? 

Can we improve these lines code with the 'general type' password  like  https://github.com/jimma/jbossws-cxf/commit/e1129a28c9824525420a7f5144f9f05fe78a79af ?
```
Password recoveredTwoWayPassword = identity.getCredential(PasswordCredential.class).getPassword();
PasswordFactory passwordFactory =PasswordFactory.getInstance(recoveredTwoWayPassword.getAlgorithm());
String expectedPassword = new String(passwordFactory.getKeySpec(passwordFactory.translate(recoveredTwoWayPassword), ClearPasswordSpec.class).getEncodedPassword());
```   
 Is it better and safe to retrieve the password object ?

Reply:
@Skyllarr  I looked at the code and there is no code to forbidden the password is the DigestPassword read from application properties. The NPE will be thrown if the `isDigest `is `true` and the password is a digest one like from the test case .

Reply:
@jimma For UsernameToken profile DigestPassword type it has to be TwoWayPassword, I explained why in this [jira comment](https://issues.redhat.com/browse/JBEAP-25738?focusedId=23186543&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23186543), I also put a comment to the code. It is also mentioned in the UsernameToken spec (https://docs.oasis-open.org/wss/v1.1/wss-v1.1-spec-pr-UsernameTokenProfile-01.htm) that the plain text password has to be recoverable:

>"Note that PasswordDigest can only be used if the plain text password (or password equivalent) is available to both the requestor and the recipient."

However the nullcheck should be added. If the recovered TwoWayPassword is null we can throw authenticationFailed.I will update this PR